hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationStats.java

```diff
@@ -93,19 +95,29 @@ void incrementCompletedMigrations() {
         totalCompletedMigrations.incrementAndGet();
     }
 
-    void recordMigrationOperationTime(long time) {
-        elapsedMigrationOperationTime.addAndGet(time);
-        totalElapsedMigrationOperationTime.addAndGet(time);
+    private void calculateElapsed(final AtomicLong elapsedTime, final AtomicLong totalElapsedTime) {
+        // This is called from each migration thread, so calculate the wall-clock time, rather than summing each
+        // individual threads execution time
+        final long newElapsed =
+                Duration.ofMillis(Clock.currentTimeMillis()).minusMillis(lastRepartitionTime.get()).toNanos();
```

Comment:
> Even though I appreciate the style, this statement allocates two new Duration objects (ofMillis and minusMillis) only to be discarded. The code is not on the hot path of operation execution, so that's not a critical issue, but it's good to be aware of allocations that produce litter. I 'd rather avoid that and write some, admittedly, more primitive code to do the math.

Ok - I've updated that.

Reply:
> That said, there is a more important issue here: Clock.currentTimeMillis() is based on System.currentTimeMillis() which is a non-monotonic clock so it can appear to jump back in time. This property makes it unsuitable for elapsed time measurements; instead we should calculate elapsed time as difference of two System.nanoTime values. So lastRepartitionTime (derived from currentTimeMillis) should not be used as start time of calculation.

I was not aware of the difference of monotonicity between `System.currentTimeMillis()` and `System.nanoTime()` - but based on that I think the only solution is another variable to store the same thing, to ensure that we have a reference for execution timing and simultaneously one for the real-world time of the last repartitioning - which seems sub-optimal.